Developer productivity workshop (#26)
Arcadia-Science/arcadia-computational-training: docs/arcadia-users-group/20221011-developer-productivity/lesson.md

- insertFinalNewline
- **Format on Save:** Search for "Format on Save" and turn it on. This will make sure, whatever code formatter you use formats the code upon saving it.
- **Default Terminal application:** Search for "Osx Exex" and type "iTerm.app".
- **Default Terminal profile:** Search for "Default Profile: Osx" and choose "bash".
- **Adding code in the PATH:** Open the command palette with Command + Shift + P. Search for "code" and choose the "Install 'code' command in PATH". This will make it so that you can open any file/directory from the terminal with `code`.

### Extensions

Before we go through this section, let's quit VS Code. Renavigate to the correct demo folder and re-open with VS Code:

```{bash}
cd ~/Desktop/arcadia-computational-training
code .
```

For this section, we'll only go through the installation instructions of a few VS Code extensions and give you a list of other extensions you can install on your own time. For downloading and installing extensions, click on the "Extensions" tab from the left tab, search for the name of the extension and hit "install". Here are the extensions:

> - Prettier: Customization with "Format Document With"

Author: I'll do demos of these, but haven't written detailed instructions for them, because I didn't have time. I think that's okay, but maybe I can write them out later

Reviewer: ya i think this is fine :)

Reviewer 2: 👍🏻

# Increasing developer productivity with IDEs and terminal customizations

In the last couple of weeks, we learned many [shell](../20220906-intro-to-shell1/lesson.md) and [git](../../workshops/20220920-intro-to-git-and-github/lesson.md) commands. This week, we'll take a step back to learn about using an integrated development environment (IDE) to develop our scripts/pipelines and customizing our terminals with shortcuts and git visualizations.

## Accessing the shell for this lesson

For this lesson, we need to have access to a Unix shell.
If you're not sure how to open a terminal on your computer, see [these instructions](https://swcarpentry.github.io/shell-novice/setup.html).

## Caveats

- You don't really have to do any of these customizations to get good at programming. But these are designed to make your life easier.

> - A good chunk of these customizations are personal. The goal is to learn _how_ to customize and to empower you to customize as you wish.

Reviewer: yes love this statement

- A good chunk of these customizations are stylistic. As a programmar, you spend a lot of time in your terminal or text editor. I think it's important to make these tools beautiful.
- I like bash because I've been using it for 10 years. This workshop will focus on using bash instead of zsh. But there are parallels between the two and learning to customize one will help you customize the other.

> - These type of tools can sometimes get RAM-intensive. So, beware if you have a laptop with <16GB of RAM.

Reviewer: can you flag in parenthetical which ones are the biggest offenders? like aliases shouldn't really cost any ram.

## Demo repository

> For demo purposes, we'll use the contents of the [AUG repository](https://github.com/Arcadia-Science/arcadia-computational-training). So, let's download the contents of it:

Reviewer: ```suggestion
For demo purposes, we'll use the contents of the [AUG repository](https://github.com/Arcadia-Science/arcadia-computational-training). So, let's download the contents of it. Our clone code assumes you have git configured to use an ssh key like we set up in the workshop. If you don't, head over to [this lesson](../../workshops/20220920-intro-to-git-and-github/lesson/#setting-up) for instructions on how to set it up.
```

```{bash}

> cd ~/Desktop

Reviewer: ```suggestion
cd ~/Desktop # or cd wherever you keep git repositories on your computer
```

git clone git@github.com:Arcadia-Science/arcadia-computational-training.git
cd arcadia-computational-training
ls
```

## IDEs

### What are IDEs?

People think of programming as mostly editing the source code of a web application or a script or a computational pipeline. But there are many other aspects of programming that happen all the time that we take for granted: debugging, testing, checking the code for errors, typos etc.

Integrated development environments or IDEs bring all these aspects of programming together and offer convenient features like autocomplete, syntax highlighting, linting, code execution etc.

IDEs are not necessary for programming, but with basic configuration, they can significantly increase coding productivity.

### Which IDE should I use?

> There are many IDEs or IDE-like text editors out there. Some are language-specific (think [IntelliJ](https://www.jetbrains.com/idea/) for Java). Some are generic like [Sublime Text](https://www.sublimehq.com/). Up until very recently, I've exclusively used Sublime Text and still love it for light-weight programming.

Reviewer: Do you think RStudio might be a better one as more people in this audience will have experience with R than Java? Not really that big of a deal if you don't feel like changing it, and the java example def doesn't have to be removed

> But at Arcadia Science, we'll be using [Visual Studio Code](https://code.visualstudio.com/) (aka VS Code). There are couple reasons for this:

Reviewer: I think this language might be too strong. Maybe saying something like "Today we'll cover Visual Studio Code" or we recommend or something like that

- It has a tight-integration with git and Github.
- It has a rich marketplace of free extensions, enabling you to configure your IDE to your heart's content.
- It works cross-platform and has a straightforward way of syncing your settings across devices.
- It is quite easy to configure because it suggests improvements on its own as needed.

> - It has access to a built-in terminal. Try hitting CTRL + ` when you install it.

Reviewer: This backtick is going to mess with the formatting on the rest of the page. I'm not sure how to escape it...

- It has tooling built-in to enable remote SSH access. So you can use an EC2 instance and navigate it through these tools.

### Installing VS Code

> Download and install VS Code by going to [this URL](https://code.visualstudio.com/download). Once installed, turn on sync across devices and login with your GitHub account.

Reviewer: is the syncing safe? like if i'm dealing with phi should i be concerned, or are there other privacy concerns?

Author: It's safe, added a comment

## Customizing VS Code

> You can just use VS Code as is. This section suggests a couple base-level customizations to make your life a little easier.

Reviewer: ```suggestion
You can use VS Code as is. This section suggests a couple base-level customizations to make your life a little easier.
```

try and avoid "just", "simple", and "obvious" 
https://docs.carpentries.org/topic_folders/communications/resources/style-guide.html#inclusivity

### Settings

**Note:** There are hundreds of settings you can customize, we'll only focus on the essentials.

Open up the settings panel by hitting Command + , or by selecting "Code" -> "Preferences" -> "Settings" from the top left corner. There are two ways to adjust settings:

1. You can search through them using the search bar
2. You can edit the underlying JSON blob that specifies the user settings.

Today, we'll mostly do #1, but if you're interested in following the second method, a simplified version of all my settings can be found in [this file](config-and-dotfiles/vscode_settings.json). So, let's get started.

> - **Auto Save:** Search for "Auto Save" in the settings menu. And adjust it to the non-default value (which is none). Mine is "afterDelay".

Reviewer: What does this do? auto save after 5 min or something? or when you delay in typing?

- **Font Size:** Search for "Font Size" and adjust it as you wish. I like 14pts font.
- **Trim Trailing Whitespace:**: Search for it and turn it on, so your code doesn't have trailing spaces or tabs.

> - **Insert Final Newline:** Search for it and turn it on. This will prevent you from receiving PR review comments on [this](https://stackoverflow.com/questions/729692/why-should-text-files-end-with-a-newline).

Reviewer: love this SO link

Reviewer 2: oooo Autosave!

- **Trim Final Newlines:** Again, turn it on. This will make sure there's only a single newline at the end of each file.

> - insertFinalNewline

Reviewer: is this a hold over?

- **Format on Save:** Search for "Format on Save" and turn it on. This will make sure, whatever code formatter you use formats the code upon saving it.

> - **Default Terminal application:** Search for "Osx Exex" and type "iTerm.app".

Reviewer: thsi only if you're on a mac or what does this do?

### Extensions

Before we go through this section, let's quit VS Code. Renavigate to the correct demo folder and re-open with VS Code:

```{bash}
cd ~/Desktop/arcadia-computational-training
code .
```

For this section, we'll only go through the installation instructions of a few VS Code extensions and give you a list of other extensions you can install on your own time. For downloading and installing extensions, click on the "Extensions" tab from the left tab, search for the name of the extension and hit "install". Here are the extensions:

- Prettier: Customization with "Format Document With"
- Markdown All in One: Opening Markdown previews with Command + Shift + V
- Excel Viewer: Opening CSV and Excel previews with Command + Shift + V
- GitHub Pull Requests and Issues
- Git Graph

#### Extension recommendations

- Python
- Pylance - for basic linting
- R
- Jupyter - for Jupyter notebooks
- Jupyter Cell Tags
- Jupyter Keymap
- Jupyter Notebook Renderers

> - Jupyter Slide Show

Reviewer: does this mean you can open and edit and run jupyter notebooks without install jupyter? how does versioning work with that?

Author: What does versioning mean in this context?

Reviewer: Oh just curious like if there is a specific version of Jupyter or Python associated with rendering notebooks in vscode

- One Dark Pro - for color schemes
- Nextflow specific [extensions](https://github.com/nf-core/vscode-extensionpack) recommended by the nf-core team

## Customizing the terminal

There are 3 main ways, we'll customize our terminals.

- Switching from zsh to bash if you haven't already.
- Installing [iTerm2](https://iterm2.com/), a better version of Apple's built-in Terminal application.
- Customizing bash with `.bash_profile`.

### Switching from zsh to bash

> Open the Terminal application, run `chsh -s /bin/bash` and quit the application. Next time you open, you should see bash. You'll see an alert that `zsh` the new Apple default when you open the terminal

Reviewer: I get this is typed up because this is your default, but im wondering what the practical differences are here or if it matters

Author: Like the practical differences between bash and zsh or something else? Just so I understand!

Reviewer: Ya the practical differences

Reviewer 2: yes I am wondering this as well, with `zsh` being the default on Mac I just switched to using that and therefore all the customizations are in my `.zshrc` file instead of `.bash_profile`. I guess I never took the time to read about the difference between the two and just went with the flow

### Installing and using iTerm2

Download and install iTerm2 by going to [this URL](https://iterm2.com/). There aren't too many ways to customize iTerm2. Or at least, I only customize the color scheme. My color scheme can be found [here](config-and-dotfiles/iterm.json). If you want to download something unique for you, there are many options [here](https://iterm2colorschemes.com/).

Once you select the color scheme you'd like to use, just do the following:

- Launch iTerm2
- Type CMD+i
- Navigate to Colors tab
- Click on Load Presets
- Click on Import
- Select your file
- Click on Load Presets and choose a color scheme

### Tweaking bash

> In general, most customizations go to the `.bash_profile` or `.bashrc` files. Here, we'll show a slightly modular approach, so we can simplify the files before they get too long and difficult to manage. Once you change the `.bash_profile` or the underlying configuration files, you need to make sure those changes are persisted. You can that by "sourcing" the `.bash_profile` with `source .bash_profile` or restarting your terminal.

Reviewer: ```suggestion
In general, most customizations go to the `.bash_profile` or `.bashrc` files. Here, we'll show a slightly modular approach, so we can simplify the files before they get too long and difficult to manage. Once you change the `.bash_profile` or the underlying configuration files, you need to make sure those changes are persistant. You can do that by "sourcing" the `.bash_profile` with `source .bash_profile` or restarting your terminal.
```

This file basically shapes the main prompt that shows up when you use bash. For it to be activated, make sure you also copy the [.bash_profile](config-and-dotfiles/.bash_profile) file. And run `source ~/.bash_profile`.

#### Using aliases for commonly used commands

Aliases are shortcuts for commands that you frequently use. Some of these provide stylistic changes. Some of these provide useful shortcuts for common git commands. You can customize this however you want, but a simplified version of my aliases can be found [here](config-and-dotfiles/.aliases).

#### Tweaking git

Git uses the `.gitconfig` file to configure the behavior of the git CLI. My simplified version of the configuration file can be found [here](config-and-dotfiles/.gitconfig). The two big components of this configuration file are:

- Colorful view for git branches, stages etc.
- Better visualization of the git history through the command-line

## What will happen in the future?

> - We got a glimpse of basic forms of linting with Pylance and basic forms of formatting with Prettier. We'll be doing a more in-depth workshop on Python and R specific linters (and ways of automating the linting process) so you can easily catch errors and have consistent styling for your code

Reviewer: should we plan this for after the conda stuff?

Author: Sure!
